From 1dea037fe703b26bd38d9d863a97645bcb1fbbb6 Mon Sep 17 00:00:00 2001
From: Motomu Utsumi <motomuman@google.com>
Date: Thu, 1 Aug 2024 21:27:12 +0900
Subject: [PATCH] Skip adding ingress discard rule to legacy VPN

Some legacy VPNs need to receive packets to VPN address via
non-VPN interface.

Bug: 352424251
Test: CsIngressDiscardRuleTest
(cherry picked from https://android-review.googlesource.com/q/commit:1b376f0fc40dd121cea2b4f523ab836c61ae5318)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:9a0a647f50dde062bb0846a87961607b4ca1c4c8)
Merged-In: Iad06dc69c5bf5ccda476e89ed9b0fc1bc63d2ebe
Change-Id: Iad06dc69c5bf5ccda476e89ed9b0fc1bc63d2ebe
---
 .../connectivityservice/CSIngressDiscardRuleTests.kt     | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/tests/unit/java/com/android/server/connectivityservice/CSIngressDiscardRuleTests.kt b/tests/unit/java/com/android/server/connectivityservice/CSIngressDiscardRuleTests.kt
index 77b06b2239..5e1beb08a8 100644
--- a/tests/unit/java/com/android/server/connectivityservice/CSIngressDiscardRuleTests.kt
+++ b/tests/unit/java/com/android/server/connectivityservice/CSIngressDiscardRuleTests.kt
@@ -337,6 +337,15 @@ class CSIngressDiscardRuleTests : CSTest() {
 
     @Test
     fun testVpnIngressDiscardRule_LegacyVpn() {
+        val nr = nr(TRANSPORT_VPN)
+        val cb = TestableNetworkCallback()
+        cm.registerNetworkCallback(nr, cb)
+        val nc = vpnNc(legacyVpn = true)
+        val lp = lp(VPN_IFNAME, IPV6_LINK_ADDRESS, LOCAL_IPV6_LINK_ADDRRESS)
+        val agent = Agent(nc = nc, lp = lp)
+        agent.connect()
+        cb.expectAvailableCallbacks(agent.network, validated = false)
+
         // IngressDiscardRule should not be added to Legacy VPN
         doTestVpnIngressDiscardRule_VpnType(TYPE_VPN_LEGACY, expectAddRule = false)
     }
-- 
2.34.1

